### PR TITLE
issue: 1080998 optimize ib_ctx class

### PR DESCRIPTION
**Pull request by rafiw**
This commit remove unneeded fields from ib_ctx class

Signed-off-by: Rafi Wiener <rafiw@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1496/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1497/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1498/ for details (Mellanox internal link).

**Comment by rafiw** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1500/ for details (Mellanox internal link).

File: src/vma/dev/ib_ctx_handler.cpp [hunk 1]
```diff
@@ -180,35 +179,22 @@ void ib_ctx_handler::set_flow_tag_capability(bool flow_tag_capability)
...
 
-ibv_port_state ib_ctx_handler::get_port_state(int port_num)
-{       
-        update_port_attr(port_num);
-        return m_ibv_port_attr.state;
+	memset(&port_attr, 0, sizeof(ibv_port_attr));
+	IF_VERBS_FAILURE(ibv_query_port(m_p_ibv_context, port_num, &port_attr)) {
+		ibch_logdbg("ibv_query_port failed on ibv device %p, port %d (errno=%d)", m_p_ibv_context, port_num, errno);
+	}ENDIF_VERBS_FAILURE;
+	return port_attr;
```
Comment (DanielLibenson) [inline src/vma/dev/ib_ctx_handler.cpp]: Please avoid using a struct (ibv_port_attr port_attr) as a return parameter.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1531/ for details (Mellanox internal link).

File: src/vma/dev/ib_ctx_handler.cpp [hunk 2]
```diff
@@ -123,22 +122,21 @@ ib_ctx_handler::ib_ctx_handler(struct ibv_context* ctx, ts_conversion_mode_t ctx
...
-		ibch_logpanic("ibv device %p pd allocation failure (ibv context %p) (errno=%d %m)", 
+		ibch_logpanic("ibv device %p pd allocation failure (ibv context %p) (errno=%d %m)",
 			    m_p_ibv_device, m_p_ibv_context, errno);
 	}
-
-	memset(&m_ibv_device_attr, 0, sizeof(m_ibv_device_attr));
-	vma_ibv_device_attr_comp_mask(m_ibv_device_attr);
-	IF_VERBS_FAILURE(vma_ibv_query_device(m_p_ibv_context, &m_ibv_device_attr)) {
-		ibch_logerr("ibv_query_device failed on ibv device %p (ibv context %p) (errno=%d %m)", 
+	m_p_ibv_device_attr = new vma_ibv_device_attr();
```
Comment (igor-ivanov) [inline src/vma/dev/ib_ctx_handler.cpp]: @rafiw why do you remove memset() for m_p_ibv_device_attr. `new()` doesn't initialize memory with zeroes by default so as a result this object can be uninitialized.

File: src/vma/dev/ib_ctx_handler.h [hunk 3]
```diff
@@ -45,52 +45,45 @@ class ib_ctx_handler : public event_handler_ibverbs
...
+	virtual void            handle_event_ibverbs_cb(void *ev_data, void *ctx);
 	inline void convert_hw_time_to_system_time(uint64_t hwtime, struct timespec* systime)
 	{
 		m_p_ctx_time_converter->convert_hw_time_to_system_time(hwtime, systime);
 	}
-
 private:
-	struct ibv_context*     m_p_ibv_context;
-	struct ibv_port_attr    m_ibv_port_attr;
+	void                    handle_event_DEVICE_FATAL();
```
Comment (igor-ivanov) [inline src/vma/dev/ib_ctx_handler.h]: @rafiw probably `handle_event_DEVICE_FATAL()` could be replaced with `handle_event_device_fatal` to improve coding style in this pull request too.

File: src/vma/dev/ib_ctx_handler.cpp (hunk 2 above)
Comment (rafiw) [inline src/vma/dev/ib_ctx_handler.cpp]: https://stackoverflow.com/questions/7546620/operator-new-initializes-memory-to-zero

File: src/vma/dev/ib_ctx_handler.h (hunk 3 above)
Comment (rafiw) [inline src/vma/dev/ib_ctx_handler.h]: will be fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1555/ for details (Mellanox internal link).

File: src/vma/dev/ib_ctx_handler.cpp (hunk 2 above)
Comment (igor-ivanov) [inline src/vma/dev/ib_ctx_handler.cpp]: where do you see that default constructor initialize memory in this link?

Comment (igor-ivanov) [inline src/vma/dev/ib_ctx_handler.cpp]: @rafiw probably you are right about `new T()`. I see good summary at link https://stackoverflow.com/posts/620402/revisions
But personally I would prefer to use `hard` memset() call.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1563/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1569/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1578/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1606/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1610/ for details (Mellanox internal link).

**Pull request merged**